### DISCUSSION TITLE
FAIR workflows

### DISCUSSION THREAD

**Discussion by dglaeser** [BAMresearch/NFDI4IngScientificWorkflowRequirements#72] [Ideas]
Here we can collect our ideas what exactly FAIR workflows mean

**Comment by joergfunger** [comment]
# Findable
* Metadataschema e.g. using [ROCrate for Workflows](https://bioschemas.org/profiles/ComputationalWorkflow/0.5-DRAFT-2020_07_21) or [this](https://www.researchobject.org/ro-crate/1.1/workflows.html)
* [Howto](https://about.workflowhub.eu/developer/how-to-make-a-workflow-ro-crate/) generate an ROcrate for workflow using python

# Accessible 
* uploading to a workflow store (also contributes to findable)
  * such as [https://workflowhub.eu/](workflowhub.eu) with [additional information on the project](https://about.workflowhub.eu/)
  * snakemake based [workflow store](https://snakemake.github.io/snakemake-workflow-catalog/) with quality guidelines [here](https://github.com/snakemake-workflows/docs)
  * a cwl, wdl and nextflow based workflow store [Dockstore](https://dockstore.org/) with [tutorials](https://anvilproject.org/news/2019/12/05/dockstore-training)
  * [nc-core](https://nf-co.re/) (nextflow workflow platform) 

# Reusable
* use a workflow management system and a standard such as [CWL](https://www.commonwl.org/) to describe the workflow including
  * versions
  * compute environments
* store provenance information when executing a e.g. using [cwlprov](https://github.com/common-workflow-language/cwlprov)

# Interoperable
* export to standardized formats for the description of a workflow (e.g. CWL)  and the provenance of a workflow execution (e.g. using CWLProv) or [wdl](https://github.com/openwdl/wdl)
  * [snakemake](https://snakemake.readthedocs.io/en/stable/executing/interoperability.html)
  * [nextflow](https://www.nextflow.io/blog/2017/nextflow-and-cwl.html) or [here](https://github.com/edgano/researchObject-Nextflow)
  * [galaxy](https://cwl.discourse.group/t/cwl-support-in-galaxy-challenges-design-choices-and-progress/553)
* use standardardized input/output formats

**Reply by dglaeser** [reply]
Sounds good to me. An attempt to start at a higher level:

- `Findable`: maybe the same as for any other (research) software applies? One wants to find software that handles a specific task, the difference here maybe just is that one explicitly looks for a workflow. I am not sure what differences this brings. Maybe this means that one would like to filter searches differently than for "ordinary" searches? This could include: 
    - workflow language (analogous to programming language?)
    - input & output file formats of the workflow (different to usual packages which have more than one use case typically?)
    - domain-specific metadata (this may also apply for usual software packages...)
- `Accessible`: the workflow description + all dependencies/data is publicly available
- `Reusable`: the workflow is self-contained (e.g. all steps are containerized or depend only on persistent packages) and/or all external data is publicly available and fetched automatically upon execution. Integration of the workflow as sub-workflow or reusing individual steps only depends on the support of such features by the chosen workflow system. 
- `Interoperability`: this could mean reusing the workflow as described above, but with a different workflow system than the one used originally. This means reusability is further promoted if the workflows can be converted into standardized descriptions that can be understood by different workflow systems. But it could also mean something along the lines that the workflow implementation relies on standard and open file formats instead of custom ones.

**Reply by dglaeser** [reply]
TODO: revisit https://direct.mit.edu/dint/article/2/1-2/108/10003/FAIR-Computational-Workflows

**Comment by dglaeser** [comment]
TODO: work collectively (in this comment) on the documentation... What do FAIR workflows mean? Let's create a bullet point list that combines the high-level vision, but also contains links to concrete tools and how they realize certain points...
